### ISSUE TITLE
Webpack style- and css- loaders in dist build

### ISSUE THREAD

**Issue by romanenko** [sstur/react-rte#32]
**Thanks for the great work! ^______^**

I get errors (`v0.1.13`) when I try to import `react-rte` using **browserify**:

```
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: Cannot find module '!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./Button.css' 
from '/../my-project-path/node_modules/react-rte/dist'
```

Is there a way to exclude webpack loaders in dist build? (I even found `module.hot` mentions in `dist.js` source code).

I downgraded to `v.0.1.10` and errors are gone (however Editor has no styles at all). I suppose something went wrong with #10

**Comment by sstur** [comment]
Thanks for tracking down this issue. I'll look into it!

**Comment by ffxsam** [comment]
Same here.

**Comment by RainFang** [comment]
@romanenko How did you browserify react-rte? I've been having issues.

`browserify -s "react-rte" -u react -u react-dom src/RichTextEditor.js > dist/react-rte.js`

is giving me SyntaxErrors: Unexpected token at DestroyableTrasform...

What did you do to get browserify to work?

**Comment by romanenko** [comment]
I use browserify with babelify.

My dependencies in `package.json` looks like:
```javascript
// ...
  "dependencies": {
    "babel-polyfill": "^6.7.4",
    "babel-preset-es2015": "^6.6.0",
    "babel-preset-react": "^6.5.0",
    "babel-preset-stage-0": "^6.5.0",
    "babelify": "^7.2.0",
    "browserify": "^13.0.0",
    "browserify-incremental": "^3.1.1",
    "react": "^15.0.1",
    "react-dom": "^15.0.1",
    "react-rte": "0.1.10"
  },
// ...
```
`.babelrc`: 
```javascript
{
  "presets": ["es2015", "stage-0", "react"]
}
```
..and I run browserify with arguments `-t [ babelify ]`

**Comment by Noxalus** [comment]
Just to say that I have the same error and I also use browserify with babelify.

**Comment by sstur** [comment]
I suspect this is because I'm using CSS Modules. We use `require('./file.css')` so any bundler needs to be aware of CSS Modules to know how to require a css file. Webpack, which I use, has a loader for this, but node, browserify and others are a different story.

Actually browserify has [css-modulesify](https://github.com/css-modules/css-modulesify) and for node there is a [require-hook](https://github.com/css-modules/css-modules-require-hook) but those solutions require you to modify your build process and/or node environment.

Ideally, what we want is `src/file.css` to be compiled to `lib/file.css.js` at build time so that any project can pull in react-rte and use it with your own bundler (or with Node).

I don't know if there is an existing tool to do that. In fact, I'm not even sure how it would work since in the browser we add a `<style>` to the document.head the first time a css filed is `require`d, but in Node, what would we do? I should check the [require-hook](https://github.com/css-modules/css-modules-require-hook) to see what they do.

I don't have an ideal solution at this point. For browserify I suggest you check out [css-modulesify](https://github.com/css-modules/css-modulesify) or pull in `dist/react-rte.js` directly. I guess the latter has some drawbacks because it will have inlined all the dependencies including React and Immutable.

I'm open to other suggestions too.

**Comment by sstur** [comment]
It looks like this might solve this issue:
https://github.com/istarkov/babel-plugin-webpack-loaders

I'll try it.

**Comment by gchudnov** [comment]
Just bumped into the same issue..

```
Cannot find module '!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./Button.css'
```
and
```
Cannot find module '!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./ButtonWrap.css'
```

Any workarounds?
Thank you!

**Comment by sstur** [comment]
I think the shortest path to a solution here is using webpack to bundle everything but the peer dependencies. I thought that's what we're already doing actually.

The more I think about it, I'm a little unclear why you'd be running into this error since your packager should be loading `react-rte/dist/react-rte.js` which is what's specified in the `main` field of package.json.

**Comment by martijnbac** [comment]
Same issue here.

`Cannot find module '!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./Dropdown.css'`

**Comment by sstur** [comment]
@martijnbac can you give me some steps to reproduce this error? I'd like to look into it some more.

**Comment by martijnbac** [comment]
I'm using [this gulpfile](https://gist.github.com/JeffreyWay/911a33403f4624753ab7). Then I import `react-rte` and run `gulp browserify`.

**Comment by jtag05** [comment]
I'm also encountering the same issue. Has there been any progress made in a browserify compatible bundle?

**Comment by gre** [comment]
same issue..

**Comment by Comaleaf** [comment]
I'm having this problem too.

@sstur The problem isn't that the package manager isn't loading from `dist/` – that is file it's loading (and I believe one work-around to this issue would be to remove the relevant `require`'s from that file).

e.g. this section from `dist/react-rte.js` starting at line 19693:

```javascript
	if(false) {
		// When the styles change, update the <style> tags
		if(!content.locals) {
			module.hot.accept("!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./Button.css", function() {
				var newContent = require("!!./../../node_modules/css-loader/index.js?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!./Button.css");
				if(typeof newContent === 'string') newContent = [[module.id, newContent, '']];
				update(newContent);
			});
		}
		// When the module is disposed, remove the <style> tags
		module.hot.dispose(function() { update(); });
	}
```

Note that the `require`'s are present, but execution will never reach them due to the `if(false)`.

browserify walks the AST of a file and resolves any `require` it finds - it does not do any static analysis to determine if execution could never reach that point (and that doesn't seem like a thing it should do).

I have been digging through the documentation and I'm not sure there is a way to fix this from configuration of browserify by selectively ignoring certain `require`s. I've no experience with Webpack so I don't know how to fix it from that end either.

As an immediate workaround you can monkey-patch the file in your node_modules, either by hand or automatically:

`sed -i -e "s/require(\"\!\!/String(\"/" node_modules/react-rte/dist/react-rte.js`

This will replace any `require("!!` with `String("` and will no longer be resolved by browserify.

If there's no obvious solution perhaps this should be part of the build step for the project. I can write a PR but it feels like quite a brittle solution.

**Comment by Comaleaf** [comment]
Ah, somebody has already made a PR at #80 that resolves this (and much more smartly) by eliminating dead code.

**Comment by sstur** [comment]
Thanks for opening this issue. I have merged #80 which should solve this. I will close this issue but please open a new one if it is not fixed in master branch.

**Comment by sstur** [comment]
By the way @Comaleaf, if you find out any way to tell Webpack to NOT add those `if (false) { ... }` blocks, please let me know!
